File: src/test/scala/com/gitlab/dhorman/cryptotrader/util/RequestLimiterSpec.scala
package com.gitlab.dhorman.cryptotrader.util

import java.util.concurrent.CountDownLatch

import com.typesafe.scalalogging.Logger
import org.scalatest.FlatSpec
import reactor.core.scala.publisher.Flux

import scala.concurrent.duration._

class RequestLimiterSpec extends FlatSpec {
  private val logger = Logger[RequestLimiterSpec]

  "RequestLimiter" should "correctly sync N requests in specified time interval" in {
    val reqLimiter = new RequestLimiter(2, 3.seconds)
    val reqCount = 10
    val additionalReqCount = 5
    val additionalReqSleepCount = Array(2000, 1000, 14000, 1000, 1000)
    val latch = new CountDownLatch(reqCount + additionalReqCount)

    def sendRequest(id: Int): Unit = {
      Flux.just(id).delaySubscription(reqLimiter.get()).subscribe(v => {
        logger.info(s"received: $v")
        latch.countDown()
      })
    }

    for (i <- 0 until reqCount) {
      sendRequest(i)
    }

    for (i <- 0 until additionalReqCount) {
      Thread.sleep(additionalReqSleepCount(i))
      sendRequest(reqCount + i)
    }

    latch.await()
  }

}

File: src/test/kotlin/com/gitlab/dhorman/cryptotrader/util/RequestLimiterSpec.kt
package com.gitlab.dhorman.cryptotrader.util

import kotlinx.coroutines.runBlocking
import mu.KotlinLogging
import org.junit.jupiter.api.Test
import reactor.core.publisher.Flux
import java.time.Duration
import java.util.concurrent.CountDownLatch


class RequestLimiterSpec {
    private val logger = KotlinLogging.logger {}

    @Test
    fun `RequestLimiter should correctly sync N requests in specified time interval`() = runBlocking {
        val reqLimiter = RequestLimiter(2, Duration.ofSeconds(3))
        val reqCount = 10
        val additionalReqCount = 5
        val additionalReqSleepCount = arrayOf<Long>(2000, 1000, 14000, 1000, 1000)
        val latch = CountDownLatch(reqCount + additionalReqCount)

        suspend fun sendRequest(id: Int) {
            Flux.just(id).delaySubscription(reqLimiter.get()).subscribe { v ->
                logger.info("received: $v")
                latch.countDown()
            }
        }

        for (i in 0 until reqCount) {
            sendRequest(i)
        }

        for (i in 0 until additionalReqCount) {
            Thread.sleep(additionalReqSleepCount[i])
            sendRequest(reqCount + i)
        }

        latch.await()
    }
}
